File: mongo-mcp-server/src/handlers/userHandlers.js
import { ObjectId } from "mongodb";
import { defaultDatabaseName } from "../config.js";
import { getDatabase } from "../mongo.js";
import { textContent } from "../response.js";

const userCollection = () => getDatabase(defaultDatabaseName).collection("user");

export const userHandlers = {
  add_user: async ({ firstName, lastName, email }) => {
    if (
      typeof firstName !== "string" ||
      typeof lastName !== "string" ||
      typeof email !== "string"
    ) {
      throw new Error("firstName, lastName, and email must be strings");
    }

    const result = await userCollection().insertOne({
      firstName,
      lastName,
      email
    });

    return textContent({
      insertedId: result.insertedId,
      firstName,
      lastName,
      email
    });
  },

  update_user: async ({ id, firstName, lastName, email }) => {
    if (typeof id !== "string" || !ObjectId.isValid(id)) {
      throw new Error("id must be a valid MongoDB ObjectId string");
    }

    const update = {};

    if (firstName !== undefined) {
      if (typeof firstName !== "string") {
        throw new Error("firstName must be a string");
      }
      update.firstName = firstName;
    }

    if (lastName !== undefined) {
      if (typeof lastName !== "string") {
        throw new Error("lastName must be a string");
      }
      update.lastName = lastName;
    }

    if (email !== undefined) {
      if (typeof email !== "string") {
        throw new Error("email must be a string");
      }
      update.email = email;
    }

    if (Object.keys(update).length === 0) {
      throw new Error("At least one of firstName, lastName, or email is required");
    }

    const result = await userCollection().updateOne(
      { _id: new ObjectId(id) },
      { $set: update }
    );

    return textContent({
      matchedCount: result.matchedCount,
      modifiedCount: result.modifiedCount,
      updatedFields: update
    });
  },

  delete_user: async ({ id }) => {
    if (typeof id !== "string" || !ObjectId.isValid(id)) {
      throw new Error("id must be a valid MongoDB ObjectId string");
    }

    const result = await userCollection().deleteOne({
      _id: new ObjectId(id)
    });

    return textContent({
      deletedCount: result.deletedCount
    });
  }
};

File: python-mcp-workspace/mongo_mcp_server/handlers/user_handlers.py
from bson import ObjectId

from ..config import DEFAULT_DATABASE_NAME
from ..mongo import get_database
from ..serializers import to_jsonable


def user_collection():
    return get_database(DEFAULT_DATABASE_NAME)["user"]


def add_user(firstName: str, lastName: str, email: str) -> dict:
    if not all(isinstance(value, str) for value in [firstName, lastName, email]):
        raise ValueError("firstName, lastName, and email must be strings")

    result = user_collection().insert_one(
        {
            "firstName": firstName,
            "lastName": lastName,
            "email": email,
        }
    )

    return to_jsonable(
        {
            "insertedId": result.inserted_id,
            "firstName": firstName,
            "lastName": lastName,
            "email": email,
        }
    )


def update_user(
    id: str,
    firstName: str | None = None,
    lastName: str | None = None,
    email: str | None = None,
) -> dict:
    if not isinstance(id, str) or not ObjectId.is_valid(id):
        raise ValueError("id must be a valid MongoDB ObjectId string")

    update = {}

    if firstName is not None:
        if not isinstance(firstName, str):
            raise ValueError("firstName must be a string")
        update["firstName"] = firstName

    if lastName is not None:
        if not isinstance(lastName, str):
            raise ValueError("lastName must be a string")
        update["lastName"] = lastName

    if email is not None:
        if not isinstance(email, str):
            raise ValueError("email must be a string")
        update["email"] = email

    if len(update) == 0:
        raise ValueError("At least one of firstName, lastName, or email is required")

    result = user_collection().update_one(
        {"_id": ObjectId(id)},
        {"$set": update},
    )

    return {
        "matchedCount": result.matched_count,
        "modifiedCount": result.modified_count,
        "updatedFields": update,
    }


def delete_user(id: str) -> dict:
    if not isinstance(id, str) or not ObjectId.is_valid(id):
        raise ValueError("id must be a valid MongoDB ObjectId string")

    result = user_collection().delete_one({"_id": ObjectId(id)})

    return {
        "deletedCount": result.deleted_count,
    }
